rfcs/0159-private-task-logs.md

```diff
@@ -0,0 +1,66 @@
+When the task is complete, they create a new S3 artifact containing the whole log, named `public/logs/live_backing.log`.
+They then modify the URL in the `public/logs/live.log` artifact (using a [special case in the `createTask` API](https://docs.taskcluster.net/docs/reference/platform/queue/api#createArtifact)) to point to the URL for the `getArtifact` method for the `public/logs/live_backing.log` artifact (`<rootUrl>/api/queue/v1/task/<taskId>/runs/<runId>/artifacts/public/logs/live_backing.log`).
+
+Reference artifacts correspond to HTTP 303 redirects, and HTTP clients do not re-authenticate when following a redirect.
+So the fetch of `live_backing.log` is performed without Taskcluster credentials, even if the initial call to `getArtifact` used credentials.
+If `live_backing.log` is always public, this is not a problem, but if it is private then this may result in additional complexity for clients wishing to access task logs.
+
+## New Log Implementation
+
+The updated convention is, briefly, to make the artifact name containing logs configurable on a per-task basis, so that users may choose a `public/` or non-public name; and to introduce a new artifact type that will avoid the issues around authentication and redirects.
+
+The convention for task-related conventions is to put information in `task.extra`, using sub-objects to group related configuration.
+In the new design, the log artiact name is found at `task.extra.logs.taskLogArtifact`, with a default value of `public/logs/live.log` for compatibility.
+Workers will be modified to consult this field before creating task log artifacts.
+UIs and other log consumers will be modified to consult this field before fetching task logs.
+
+To support live logging, a new queue artifact type, "LiveLog", is added.
+This type is similar to the existing Reference type, in that it contains a URL to which it redirects.
+However, its URL cannot be modified, but it can be replaced with an artifact of another type.
+
+Workers create a LiveLog artifact when a task begins, and replace that with a "normal" artifact (such as S3) when the task completes.
+The URL in the LiveLog artifact should contain enough entropy that it cannot be guessed, preventing anyone who cannot read the LiveLog artifact from connecting to the worker to download logs.
+
+# Implementation
+
+The following changes will be required (this section will be updated with progress):
+
+ * Documentation for new convention
+ * Implementation of new LiveLog artifact type in queue
+ * Implementation of `task.extra.logs.taskLogArtifact`, fetching logs with credentials in Taskcluster UI
+ * Implementation of updated log handling in workers
+   * Generic-Worker
+   * Docker-Worker
+   * Scriptworker
+ * Implementation in other log consumers; known consumers:
+   * Treeherder (NOTE: Treeherder currently fetches `public/logs/live_backing.log` directly)
```

Comment:
What timeline are you looking at to getting to this part?
What changes will be needed?

Reply:
Timeline isn't decided yet -- it will depend where this overall project falls in the TC team's priorities, and whether folks outside of the team can work on it.  I suspect the TH changes would be relatively minor, too.

Reply:
I think the changes would be:
 * look for the artifact name in task.extra, defaulting to `public/tasks/live.log`
 * stop referring directly to `live_backing.log`

The second step may need to be timed to be close to deployment of the LiveLog artifact type, as otherwise it means slower log processing as the requests follow a redirect from `live.log` to `live_backing.log`.